pkg/util/cdcutil/cdc.go

```diff
@@ -37,120 +40,227 @@ const (
+		return invalidTs, nil
+	case "running", "warning", "normal", "stopped":
+		cts, err := c.fetchCheckpointTSFromStatus(ctx, cf)
 		if err != nil {
-			return nil, errors.Trace(err)
+			return 0, err
 		}
-		for _, kv := range resp.Kvs {
-			// example: /tidb/cdc/changefeed/info/<changefeedID>
-			k := kv.Key[len(CDCPrefixV61):]
-			if len(k) == 0 {
-				continue
-			}
-			if !isActiveCDCChangefeed(kv.Value) {
-				continue
-			}
-
-			nameSet["<nil>"] = append(nameSet["<nil>"], string(k))
+		return mathutil.Max(cts, info.Start), nil
+	default:
+		// This changefeed may be noise, ignore it.
+		log.Warn("Ignoring invalid changefeed.", zap.Any("changefeed", cf), zap.String("state", info.State))
+		return invalidTs, nil
+	}
+}
+
+func (c checkCDCClient) getIncompatible(ctx context.Context, safeTS uint64) (*CDCNameSet, error) {
+	changefeeds := make([]changefeed, 0)
+	if err := c.loadChangefeeds(ctx, &changefeeds); err != nil {
+		return nil, err
+	}
+
+	nameset := new(CDCNameSet)
+	for _, cf := range changefeeds {
+		cts, err := c.checkpointTSFor(ctx, cf)
+		if err != nil {
+			return nil, errors.Annotatef(err, "failed to check changefeed %v", cf)
+		}
+		if cts < safeTS {
+			log.Info("Found incompatible changefeed.", zap.Any("changefeed", cf), zap.Uint64("checkpoint-ts", cts), zap.Uint64("safe-ts", safeTS))
+			nameset.save(cf)
 		}
 	}
 
-	return &CDCNameSet{nameSet}, nil
+	return nameset, nil
 }
 
-type onlyState struct {
-	State string `json:"state"`
+// CDCNameSet saves CDC changefeed's information.
+// nameSet maps `cluster/namespace` to `changefeed`s
+type CDCNameSet struct {
+	changefeeds map[string][]string
 }
 
-func isActiveCDCChangefeed(jsonBytes []byte) bool {
-	s := onlyState{}
-	err := json.Unmarshal(jsonBytes, &s)
-	if err != nil {
-		// maybe a compatible issue, skip this key
-		log.L().Error("unmarshal etcd value failed when check CDC changefeed, will skip this key",
-			zap.ByteString("value", jsonBytes),
-			zap.Error(err))
-		return false
+func (s *CDCNameSet) save(cf changefeed) {
+	if s.changefeeds == nil {
+		s.changefeeds = map[string][]string{}
 	}
-	switch s.State {
-	// https://docs.pingcap.com/zh/tidb/stable/ticdc-changefeed-overview
-	case "normal", "stopped", "error", "warning":
-		return true
-	default:
-		return false
+
+	switch cf.KeyVersion {
+	case legacy:
+		s.changefeeds["<nil>"] = append(s.changefeeds["<nil>"], cf.ID)
+	case namespaced:
+		k := path.Join(cf.Cluster, cf.Namespace)
+		s.changefeeds[k] = append(s.changefeeds[k], cf.ID)
 	}
 }
+
+// Empty that the nameSet is empty means no changefeed exists.
+func (s *CDCNameSet) Empty() bool {
+	return len(s.changefeeds) == 0
+}
+
+// MessageToUser convert the map `nameSet` to a readable message to user.
+func (s *CDCNameSet) MessageToUser() string {
+	var changefeedMsgBuf strings.Builder
+	changefeedMsgBuf.WriteString("found CDC changefeed(s): ")
+	for clusterID, changefeedID := range s.changefeeds {
+		changefeedMsgBuf.WriteString("cluster/namespace: ")
```

Comment:
namespace is a part of changefeed ID

Reply:
Here I'm going to try keep the old output format. Some of test cases relies on it.

Reply:
Here I'm going to try keep the old output format. Some of test cases rely on it.